### PR TITLE
Manhwabuddy

### LINKED ISSUE #4809: ManhwaBuddy source
### Source name

ManhwaBuddy

### Source link

https://manhwabuddy.com/

### Source language

English

### Other details

18+/NSFW

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by dngonz**
Closes #4809

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.Request
+import org.jsoup.nodes.Document
+import org.jsoup.nodes.Element
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class ManhwaBuddy : ParsedHttpSource() {
+    override val baseUrl = "https://manhwabuddy.com/"
```
Comment (vetleledaal) [inline src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt]: Remove trailing slash

File: src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    override val lang = "en"
+    override val name = "ManhwaBuddy"
+    override val supportsLatest = true
+
+    override fun chapterFromElement(element: Element): SChapter {
+        return SChapter.create().apply {
+            setUrlWithoutDomain(element.attr("abs:href"))
+            name = element.selectFirst(".chapter-name")!!.text()
+            date_upload = element.selectFirst(".ct-update")?.text()?.let { parseDate(it) }!!
```
Comment (vetleledaal) [inline src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt]: `date_upload` isn't critical, you should not use a non-null assert here.

File: src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    override fun imageUrlParse(document: Document): String {
+        throw UnsupportedOperationException()
+    }
+
+    override fun latestUpdatesFromElement(element: Element): SManga {
+        return SManga.create().apply {
+            setUrlWithoutDomain(element.selectFirst("a")!!.attr("abs:href"))
+            title = element.selectFirst("h4")!!.text()
+            thumbnail_url = element.selectFirst("img")!!.attr("src")
```
Comment (vetleledaal) [inline src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt]: Not critical, remove non-null assert.

Comment also applies to wherever you have set `thumbnail_url` throughout this file.

File: src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    private val dateFormat by lazy {
+        SimpleDateFormat("dd MMMM yyyy", Locale.ENGLISH)
+    }
+
+    override fun chapterListSelector(): String {
+        return ".chapter-list a"
+    }
```
Comment (vetleledaal) [inline src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt]: Functions that return a string can generally be simplified.
```suggestion
    override fun chapterListSelector(): String = ".chapter-list a"
```

Comment also applies to similar patterns used throughout this file.

File: src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    override fun mangaDetailsParse(document: Document): SManga {
+        return SManga.create().apply {
+            val info = document.selectFirst(".main-info-right")!!
+            author = info.selectFirst("li:contains(Author(s)) a")?.text()
+            status = when (info.selectFirst("li:contains(Status) span")?.text()) {
+                "Ongoing" -> SManga.ONGOING
+                "Complete" -> SManga.COMPLETED
+                else -> SManga.UNKNOWN
+            }
+            artist = info.selectFirst("li:contains(Artist(s)) a")?.text()
```
Comment (vetleledaal) [inline src/en/manhwabuddy/src/eu/kanade/tachiyomi/extension/en/manhwabuddy/ManhwaBuddy.kt]: Does Jsoup properly tokenize the selector here? To be on the safe side, consider:
```css
li:contains(Artist) a
```

**Review by vetleledaal — CHANGES_REQUESTED** [review]

**Comment by vetleledaal** [comment]
For future reference, you don't need a corresponding issue for each PR. If there are no related issues it means you have linked all related issues (none).

**Comment by dngonz** [comment]
Hi @vetleledaal thanks for the review, changes are already applied, also added a new change just to improve the description.

File: src/en/manhwabuddy/res/output-onlinepngtools.png
Comment (AwkwardPeak7) [inline src/en/manhwabuddy/res/output-onlinepngtools.png:1]: remove

**Review by vetleledaal — APPROVED** [review]

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**